### PR TITLE
Emit DWARF info for local variables and function parameters

### PR DESCRIPTION
**Pull request by mshinwell**
This is a port of some of my previous work to current flambda-backend.  It still needs tidying up and fixing, but should be in a reasonable shape for basic support soon.

Includes #1651 and #1653.

### CODE REVIEW

File: ocaml/typing/ident.ml [hunk 1]
```diff
@@ -154,6 +154,15 @@ let is_instance = function
...
   | _ -> false
 
+let is_optional_parameter t =
+  match t with
+  | Local { name; _ }
+  | Scoped { name; _ } ->
+    if String.length name > 5 && String.equal (String.sub name 0 5) "*opt*"
+    then Some (String.sub name 5 (String.length name - 5))
+    else None
+  | Global _ | Predef _ | Instance _ -> None
```
Comment (lpw25) [inline ocaml/typing/ident.ml:164]: This isn't really about identifiers, can you move this logic elsewhere? Perhaps in `backend_var`? Also, maybe we should work out how to get rid of these terrible names.

**Comment by mshinwell** [comment]
I think `-g -gno-upstream-dwarf` is needed to try this.  We are mainly targeting lldb, use this branch: https://github.com/ocaml-flambda/llvm-project

If using macOS you need to run `dsymutil` on the executable.

**Comment by mshinwell** [comment]
On Linux this currently needs `--disable-function-sections` to `configure`

**Comment by mshinwell** [comment]
@poechsel the head commit here is ready for review.  The function sections support still needs implementing, I'll do that soon.

File: backend/debug/dwarf/dwarf_flags/dwarf_flags.ml [hunk 2]
```diff
@@ -72,7 +72,7 @@ let use_g () =
   then use_g1 ()
   else current_debug_settings := bytecode_g
 
-let restrict_to_upstream_dwarf = ref true
+let restrict_to_upstream_dwarf = ref false (* XXX *)
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply backend/debug/dwarf/dwarf_flags/dwarf_flags.ml]: That was for debugging, has been removed

File: backend/debug/dwarf/dwarf_ocaml/dwarf_concrete_instances.ml [hunk 3]
```diff
@@ -25,12 +25,35 @@ type fundecl =
...
+let remove_double_underscores s =
+  let len = String.length s in
+  let buf = Buffer.create len in
+  let skip = ref false in
+  let rec loop i =
+    if i < len
+    then (
+      let c = String.get s i in
+      if c = '.' then skip := true;
+      if (not !skip) && c = '_' && i + 1 < len && String.get s (i + 1) = '_'
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply backend/debug/dwarf/dwarf_ocaml/dwarf_concrete_instances.ml:37]: We are, but I also think using the unsafe function is unlikely to have a noticeable effect on performance.  We should err on the side of caution in the compiler

File: backend/debug/dwarf/dwarf_ocaml/dwarf.ml [hunk 4]
```diff
@@ -39,16 +40,36 @@ let create ~sourcefile ~unit_name ~asm_directives ~get_file_id ~code_begin
...
+    Proto_die.create ~parent:(Some compilation_unit_proto_die) ~tag:Base_type
+      ~attribute_values:
+        [ DAH.create_name "ocaml_value";
+          DAH.create_encoding ~encoding:Encoding_attribute.signed;
+          DAH.create_byte_size_exn ~byte_size:Arch.size_addr ]
+      ()
+  in
+  let start_of_code_symbol =
+    Cmm_helpers.make_symbol "code_begin" |> Asm_symbol.create
+    (* Dwarf_name_laundry.mangle_symbol Text (Symbol.of_global_linkage
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply backend/debug/dwarf/dwarf_ocaml/dwarf.ml:53]: Fixed

**Comment by mshinwell** [comment]
Ten years ago, we started writing a register availability analysis, so we could see the values of variables in the debugger...

**Pull request merged**